### PR TITLE
WIP: New instances

### PR DESCRIPTION
**Pull request by arthurdouillard**
~Still work in progress.~

### CODE REVIEW

File: clloader/scenarios/instance_incremental.py [hunk 1]
```diff
@@ -27,10 +28,11 @@ class InstanceIncremental(_BaseCLLoader):
     def __init__(
         self,
         cl_dataset: _ContinuumDataset,
-        nb_tasks: int,
+        nb_tasks: int = 0,
```
Comment (TLESORT) [inline clloader/scenarios/instance_incremental.py:31]: Ca veut dire quoi en instance incremental nb_tasks=0? pour moi c'est un argument obligatoire dans ce scenario

File: clloader/scenarios/instance_incremental.py [hunk 2]
```diff
@@ -40,14 +42,35 @@ def __init__(
...
     def _setup(self, nb_tasks: int) -> int:
-        x, y, _ = self.cl_dataset.init(train=self.train)
+        x, y, t = self.cl_dataset.init(train=self.train)
 
-        # TODO: need to add seed + randomstate to ensure reproducibility
-        task_ids = np.random.randint(nb_tasks, size=len(y))
+        if t is None and nb_tasks <= 0:
+            raise ValueError(f"You need to specify a number of tasks > 0, not {nb_tasks}.")
+        elif t is None:
+            task_ids = self._random_state.randint(nb_tasks, size=len(y))
```
Comment (TLESORT) [inline clloader/scenarios/instance_incremental.py:55]: je pense que la definition des task_ids devrait provenir d'une fonction "get_task_ids()". Comme ça ce sera facile a changer pour les classes filles

File: clloader/scenarios/instance_incremental.py [hunk 3]
```diff
@@ -40,14 +42,35 @@ def __init__(
...
+            self.dataset = (x, y, task_ids)
+        else:
+            default_nb_tasks = len(np.unique(t))
+            if nb_tasks > 0 and default_nb_tasks > nb_tasks:
+                warnings.warn(
+                    f"The default number of task ({default_nb_tasks} is lower than"
+                    f" the one asked ({nb_tasks}), some tasks will be removed."
+                )
+                indexes = np.where(t <= nb_tasks - 1)[0]
+                x, y, t = x[indexes], y[indexes], t[indexes]
```
Comment (TLESORT) [inline clloader/scenarios/instance_incremental.py:65]: C'est pour le cas ou les task_ids sont donnés par le user? mets un ptit commentaire ptet

File: clloader/task_set.py [hunk 4]
```diff
@@ -1,11 +1,12 @@
 from typing import Callable, List, Tuple, Union
 
-import torch
 import numpy as np
-from clloader.viz import plot
+import torch
```
Comment (TLESORT) [inline clloader/task_set.py:4]: unused I think

File: tests/test_instance_incremental.py [hunk 5]
```diff
@@ -23,15 +25,79 @@ def numpy_data():
...
+
+    return (x_train, y_train.astype(int), t_train), (x_test, y_test.astype(int), t_test)
+
+
+@pytest.mark.parametrize("nb_tasks,nb_tasks_gt,catch", [
+    (2, 2, False),
+    (6, None, True),
+    (0, 3, False),
+])
+def test_instance_default_nb_tasks(numpy_data_per_task, nb_tasks, nb_tasks_gt, catch):
```
Comment (TLESORT) [inline tests/test_instance_incremental.py:76]: We should maybe add a small comment at the beginning of a test to explain what it does

File: clloader/datasets/core50.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+class CORe50(_ContinuumDataset):
+
+    def __init__(
+        self, folder: str, train_image_ids: Union[str, List[str]], download: bool = True, **kwargs
+    ):
+        super().__init__(download=download, **kwargs)
+
+        self.folder = folder
+        self.train_image_ids = train_image_ids
```
[the remark this answers is not in the document]
Comment (arthurdouillard) [reply clloader/datasets/core50.py:18]: Si train_ids, si l'image n'appartient pas à ce subset, c'est qu'elle appartient au test set.

File: clloader/datasets/core50.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+                    else:
+                        t.append(0)
+
+                    has_images = True
+            if has_images:
+                domain_counter += 1
+
+        x = np.array(x)
+        y = np.array(y)
+        t = np.array(t)
```
[the remark this answers is not in the document]
Comment (arthurdouillard) [reply clloader/datasets/core50.py:75]: Done.

File: clloader/scenarios/base.py [hunk 8]
```diff
@@ -27,8 +27,11 @@ def __init__(
         if common_transformations is None:
             common_transformations = self.cl_dataset.transformations
 
-        self.train_trsf = transforms.Compose(train_transformations + common_transformations)
-        self.test_trsf = transforms.Compose(common_transformations)
+        if len(common_transformations) == 0:
+            self.train_trsf, self.test_trsf = None, None
+        else:
+            self.train_trsf = transforms.Compose(train_transformations + common_transformations)
+            self.test_trsf = transforms.Compose(common_transformations)
```
[the remark this answers is not in the document]
Comment (arthurdouillard) [reply clloader/scenarios/base.py:34]: To handle the case w/o any transfo (i.e. nlp).

I agree it's not amazing, but as you said the difference between those two will be removed.

File: clloader/scenarios/instance_incremental.py (hunk 1 above)
Comment (arthurdouillard) [reply clloader/scenarios/instance_incremental.py:31]: If we want to use the default number of tasks. Like for CORe50, there are 7 tasks by default, if the user left the `nb_tasks` unspecified (here `0`, but it could be `-1`) we have 7 tasks.

However the user could provide a number lower than 7, and we would truncate those tasks.

Note that if the desired `nb_tasks` is above the one designed by the dataset, we raise an error.

File: clloader/scenarios/instance_incremental.py (hunk 2 above)
Comment (arthurdouillard) [reply clloader/scenarios/instance_incremental.py:55]: Do we want to create a new class inheriting from `InstanceIncremental`? I'm sure why.

Or you simply want that the `_setup` method just call directly a new method `get_task_ids`?

File: clloader/scenarios/instance_incremental.py (hunk 1 above)
Comment (TLESORT) [reply clloader/scenarios/instance_incremental.py:29]: ok

File: clloader/scenarios/instance_incremental.py (hunk 2 above)
Comment (TLESORT) [reply clloader/scenarios/instance_incremental.py:53]: I suggested that _setup method call a new method get_task_ids instead of computing the task label inside _setup function

File: clloader/datasets/core50.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+                        continue
+                    elif not train and image_id in train_images_ids:
+                        continue
+
+                    x.append(os.path.join(object_folder, path))
+                    y.append(object_id)
+                    if train:
+                        t.append(domain_counter)
+                    else:
+                        t.append(0)
```
Comment (TLESORT) [inline clloader/datasets/core50.py:100]: In the NI scenario, it is possible to identify a specific test image to a specific task, would it be usefull to do it?

**Comment by TLESORT** [comment]
You should merge the master on your branch to solve conflict I guess

File: clloader/datasets/core50.py (hunk 9 above)
Comment (arthurdouillard) [reply clloader/datasets/core50.py:100]: Sure, but what about this case where train and test domains are different?

Train may have the domain ids [0, 1, 2, 3, 4, 5, 6]; and test [0, 1, 2]. But train#0 and test#0 are two different domains.

To avoid thinking it was the same domain, I prefered setting a dummy one for all.

Do you see a better solution?

File: clloader/scenarios/instance_incremental.py (hunk 2 above)
Comment (arthurdouillard) [reply clloader/scenarios/instance_incremental.py:53]: It doesn't simply create task ids tho. In the case of asking less tasks than the default one, we also truncate x, and y.
